### PR TITLE
StandardMetric Pre-Aggregation with all expected dimensions

### PR DESCRIPTION
**Pull request by cijothomas**
Addressing #1233 

TODO:
Tests
PerformanceBucket dimension
Perf Tests

Will finalize the default dimension limits after more perf/cost research.

### CODE REVIEW

File: BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/DependencyMetricsExtractor.cs [hunk 1]
```diff
@@ -128,162 +156,30 @@ public void ExtractMetrics(ITelemetry fromItem, out bool isItemProcessed)
...
-                dependencyCallMetric.TryGetDataSeries(
-                                                    out seriesToTrack,
-                                                    MetricTerms.Autocollection.Metric.DependencyCallDuration.Id,
-                                                    dependencySuccessString,
-                                                    MetricTerms.Autocollection.DependencyCall.TypeNames.Other);
-            }
-            else
+            int i = 0;
+            string[] dimValues = new string[this.dimensionExtractors.Count];
+            foreach (var dim in this.dimensionExtractors)
```
Comment (Dmitry-Matveev) [inline BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/DependencyMetricsExtractor.cs]: dimensionExtractors is a list, I thin you can simplify and join counter `i` and list traversing into a single for loop:
```
for (int i=0; i < dimValues.Length; i++ )
{
   dimValues[i] = this.dimensionExtractors[i].ExtractDimension(dependencyCall);
}
```

It also may end up more performant (doing this for Telemetry Initializers saved me on Iterator as per profile traces).

File: BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/MetricTerms.cs [hunk 2]
```diff
@@ -75,11 +76,23 @@ public static class PropertyNames
...
+                    public const string CloudRoleInstance = "cloud/roleInstance";
+                    public const string CloudRoleName = "cloud/roleName";
+                    public const string Synthetic = "operation/synthetic";
                 }
 
-                public static class TypeNames
+                public static class PropertyValues
                 {
-                    public const string Other = "Other";
+                    public const string DimensionCapFallbackValue = "DIMENSION-CAPPED";
```
Comment (Dmitry-Matveev) [inline BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/MetricTerms.cs]: Used to be DIMENSION_CAPPED, now DIMENSION-CAPPED. Is underscore to hyphen change intentional?

File: BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/RequestExtractor/RequestIdDimensionExtractor.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation.Metrics
+{
+    using System;
+    using Microsoft.ApplicationInsights.Channel;    
+
+    internal class RequestIdDimensionExtractor : IDimensionExtractor
```
Comment (Dmitry-Matveev) [inline BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/RequestExtractor/RequestIdDimensionExtractor.cs]: RequestId & DepednencyId are often associated with RequestTelemetry.Id and DependencyTelemetry.Id - should we consider slightly different name for those methods in extractors?

File: BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/RequestMetricsExtractor.cs [hunk 4]
```diff
@@ -13,36 +14,99 @@
     /// </summary>
     internal class RequestMetricsExtractor : ISpecificAutocollectedMetricsExtractor
     {
-        private MetricSeries responseTimeSuccessSeries;
-        private MetricSeries responseTimeFailureSeries;
+        /// <summary>
+        /// The default value for the <see cref="MaxResponseCodeToDiscover"/> property.        
+        /// </summary>
+        public const int MaxResponseCodeToDiscoverDefault = 10;
```
Comment (Dmitry-Matveev) [inline BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/RequestMetricsExtractor.cs]: We do not do subcodes, right? I'd still bump it a bit as [per this list](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes), only legitimate/often 4xx and 5xx is already at the limit of 10.

File: BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/MetricTerms.cs (hunk 2 above)
Comment (cijothomas) [reply BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/MetricTerms.cs]: Its insignificant. There are no backend dependency on this value.
Once we commit to this string, will inform backend teams on this - as of now only AppMap will rely on pre-aggs so it should be taught to be aware of dim caps.
For metric explorer/other metric experiences - Dimension=capped is just a string.

File: BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/DependencyExtractor/IdDimensionExtractor.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    internal class IdDimensionExtractor : IDimensionExtractor
+    {
+        public int MaxValues { get; set; } = 1;
+        public string DefaultValue { get; set; } = MetricTerms.Autocollection.Metric.DependencyCallDuration.Id;
+
+        public string Name { get; set; } = MetricDimensionNames.TelemetryContext.Property(MetricTerms.Autocollection.MetricId.Moniker.Key);
+
+        public string ExtractDimension(ITelemetry item)
+        {
+            return MetricTerms.Autocollection.Metric.DependencyCallDuration.Id;
```
[the remark this answers is not in the document]
Comment (cijothomas) [reply BASE/src/Microsoft.ApplicationInsights/Metrics/Implementation/AutocollectedMetricsExtraction/DependencyExtractor/IdDimensionExtractor.cs]: yup. Default is never going to be used as the ExtractDimension always return a constant.

**Comment by TimothyMothra** [comment]
>  Metrics\Implementation\AutocollectedMetricsExtraction\DependencyMetricsExtractor.cs(37,43): error SA1306: Field 'DimensionExtractors' should begin with lower-case letter [E:\A\_work\31\s\BASE\src\Microsoft.ApplicationInsights\Microsoft.ApplicationInsights.csproj]

File: BASE/Test/Microsoft.ApplicationInsights.Test/Shared/Metrics/Extensibility/AutocollectedMetricsExtraction/AutocollectedMetricsExtractorTest.cs [hunk 6]
```diff
@@ -94,72 +250,202 @@ public void Request_CorrectlyExtractsMetric()
...
+                var metricCollection = serverResponseMetric.Where(
+                (tel) => (tel as MetricTelemetry).Properties["Request.Success"] == success[i].ToString());
+                int expectedCount = totalTimeSeries / success.Length;
+                Assert.AreEqual(expectedCount, metricCollection.Count());
+                ValidateAllMetric(metricCollection);
+            }
 
-            var t = new SortedList<string, MetricTelemetry>();
+            // Validate synthetic dimension
+            for (int i = 0; i < success.Length; i++)
```
Comment (Dmitry-Matveev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/Shared/Metrics/Extensibility/AutocollectedMetricsExtraction/AutocollectedMetricsExtractorTest.cs]: Synthetic.Length

File: BASE/Test/Microsoft.ApplicationInsights.Test/Shared/Metrics/Extensibility/AutocollectedMetricsExtraction/AutocollectedMetricsExtractorTest.cs [hunk 7]
```diff
@@ -379,319 +661,178 @@ public void Dependency_CanSetMaxDependencyTypesToDiscoverAfterInitialization()
...
-                Assert.AreEqual(50, metric.Sum);
+                var metricCollection = depDurationMetric.Where(
+                (tel) => (tel as MetricTelemetry).Properties["Dependency.Success"] == success[i].ToString());
+                int expectedCount = totalTimeSeries / success.Length;
+                Assert.AreEqual(expectedCount, metricCollection.Count());
+                ValidateAllMetric(metricCollection);
             }
 
+            // Validate synthetic dimension
+            for (int i = 0; i < success.Length; i++)
```
Comment (Dmitry-Matveev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/Shared/Metrics/Extensibility/AutocollectedMetricsExtraction/AutocollectedMetricsExtractorTest.cs]: Same here

File: BASE/Test/Microsoft.ApplicationInsights.Test/Shared/Metrics/Extensibility/AutocollectedMetricsExtraction/DependencyTypeDimesnionExtractorTest.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation.Metrics
+{
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    [TestClass]
+    public class DependencyTypeDimesnionExtractorTest
+    {
+        [TestMethod]
+        public void NullTarget()
```
Comment (Dmitry-Matveev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/Shared/Metrics/Extensibility/AutocollectedMetricsExtraction/DependencyTypeDimesnionExtractorTest.cs]: Type

**Pull request merged**